pkg/domain/extension.go

```diff
@@ -390,87 +361,701 @@ func (e *ErrExtensionCredentialsNotFound) Error() string {
+			if len(service.Credentials) > 0 || service.AuthRequired {
+				for _, credential := range service.Credentials {
+					accessDesc := ExtensionAccessDescriptor{
+						Extension:   *e,
+						Service:     *service,
+						Endpoint:    *endpoint,
+						Credentials: credential,
+					}
+					result = append(result, &accessDesc)
+				}
+			} else {
+				accessDesc := ExtensionAccessDescriptor{
+					Extension:   *e,
+					Service:     *service,
+					Endpoint:    *endpoint,
+					Credentials: nil,
+				}
+				result = append(result, &accessDesc)
+			}
+		}
+	}
+
+	return result
+}
+
+// FindServices returns services belonging to the extension that match the query.
+func (e *Extension) FindServices(query *ExtensionQuery) (map[string]*ExtensionService, error) {
+	result := make(map[string]*ExtensionService)
+
+	addIfMatch := func(service *ExtensionService) {
+		if query.ServiceID != "" && query.ServiceID != service.ID {
+			return
+		}
+		if query.ServiceResource != "" && query.ServiceResource != service.Resource {
+			return
+		}
+		if query.ServiceCategory != "" && query.ServiceCategory != service.Category {
+			return
+		}
+
+		endpoints, err := service.FindEndpoints(e.Zone, query)
+		if err == nil {
+			credentials, err := service.FindCredentials(query)
+			if err == nil {
+				// Create a copy of the service but containing only the endpoints and credentials that match the query
+				serviceCopy := *service
+				serviceCopy.Endpoints = endpoints
+				serviceCopy.Credentials = credentials
+				result[service.ID] = &serviceCopy
+			}
+		}
+	}
+
+	for _, service := range e.ListServices() {
+		addIfMatch(service)
+	}
+
+	return result, nil
+}
+
+// SetCreated sets the created time of the extension and its services, credentials, endpoints.
+func (e *Extension) SetCreated(ctx context.Context) {
+	e.Created = time.Now()
+	e.Updated = e.Created
+
+	for _, service := range e.ListServices() {
+		service.SetCreated(e.Created)
+	}
+}
+
+// simple unique extension ID generator
+func (e *Extension) generateExtensionID(ctx context.Context, store ExtensionStore) string {
+	prefix := e.Product
+	if prefix != "" {
+		prefix = prefix + "-"
+	}
+	for {
+		ID := prefix + rand.String(8)
+		if _, err := store.GetExtension(ctx, ID); err != nil {
+			return ID
+		}
+	}
+}
+
+// simple unique extension service ID generator
+func (e *Extension) generateExtensionServiceID(service *ExtensionService) string {
+	prefix := service.Resource
+	if prefix == "" && e.Product != "" {
+		prefix = e.Product + "-service"
+	}
+	if prefix != "" {
+		prefix = prefix + "-"
+	}
+	for {
+		ID := prefix + rand.String(8)
+		if e.Services[ID] == nil {
+			return ID
+		}
+	}
+}
+
+// EnsureIDs sets the ID of the service and its credentials when ID is empty.
+func (es *ExtensionService) EnsureIDs(ext *Extension) {
+	if es.ID == "" {
+		es.ID = ext.generateExtensionServiceID(es)
+	}
+	for _, credentials := range es.ListCredentials() {
+		credentials.EnsureID(es)
```

Comment:
I still think this is redundant, especially now that you clarified that the credential IDs are pre-ensured before they are added to the service.Credentials map.

Reply:
you're right, I have removed that.